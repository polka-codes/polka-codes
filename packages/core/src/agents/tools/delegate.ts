import { z } from 'zod'

import { type AgentInfo, agentTool } from '../types'

export const delegateTool = (agents: AgentInfo[]) => {
  const agentInputSchemas = agents.map((agent) =>
    z.object({
      agentName: z.literal(agent.name),
      task: z.string().describe('The task to be completed by the target agent'),
      ...(agent.contextSchema ? { context: agent.contextSchema.describe('The context information for the task') } : {}),
    }),
  )

  let inputSchema: z.ZodSchema
  if (agents.length === 0) {
    throw new Error('At least one agent is required to delegate a task')
  }
  if (agents.length === 1) {
    inputSchema = agentInputSchemas[0]
  } else {
    inputSchema = z.union(agentInputSchemas as any)
  }

  return agentTool({
    id: 'delegate',
    description: 'Delegate a task to another agent to execute and receive the result back',
    inputSchema: inputSchema.describe('The input for the task'),
    outputSchema: z.object({
      status: z.enum(['success', 'failure']).describe('The status of the task'),
      output: z.string().describe('The output of the task'),
    }),
    execute: async (input, polka) => {
      try {
        const result = await polka.startTask(input.agentName, input.task, input.context)
        if (typeof result === 'string') {
          return {
            status: 'success' as const,
            output: result,
          }
        }
        return {
          status: 'failure' as const,
          output: result.error,
        }
      } catch (error: any) {
        return {
          status: 'failure' as const,
          output: 'message' in error ? error.message : (error?.toString() ?? JSON.stringify(error)),
        }
      }
    },
  })
}
