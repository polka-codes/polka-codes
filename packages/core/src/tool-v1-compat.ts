// generated by polka.codes
import { z } from 'zod'
import type { FullToolInfo, FullToolInfoV2, ToolParameter } from './tool'

function zodSchemaToParameters(schema: z.ZodObject<any>): ToolParameter[] {
  const parameters: ToolParameter[] = []
  const { shape } = schema

  for (const name in shape) {
    const def: z.ZodTypeAny = shape[name]
    const isOptional = def.safeParse(undefined).success

    const description = def.description || ''

    const param: ToolParameter = {
      name,
      description,
      required: !isOptional,
    }

    const usageValue = def.meta()?.usageValue as string | undefined
    if (usageValue) {
      param.usageValue = usageValue
    }

    if (def instanceof z.ZodObject) {
      param.children = zodSchemaToParameters(def)
    } else if (def instanceof z.ZodArray) {
      param.allowMultiple = true
      const element = def.element
      if (element instanceof z.ZodObject) {
        param.children = zodSchemaToParameters(element)
      }
    }

    parameters.push(param)
  }

  return parameters
}

export function toToolInfoV1(tool: FullToolInfoV2): FullToolInfo {
  const { parameters: zodSchema, ...rest } = tool

  const v1Parameters = zodSchemaToParameters(zodSchema)

  return {
    ...rest,
    parameters: v1Parameters,
  }
}
