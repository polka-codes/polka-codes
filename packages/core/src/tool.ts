export type ToolParameter = {
  name: string
  description: string
  required: boolean
  usageValue?: string
  allowMultiple?: boolean
  children?: ToolParameter[]
}

export type ToolParameterValue = string | { [key: string]: ToolParameterValue } | ToolParameterValue[]

export type ToolExample = {
  description: string
  parameters: { name: string; value: ToolParameterValue }[]
}

export enum PermissionLevel {
  // This tool is completely safe
  None = 0,
  // This tool can read files
  Read = 1,
  // This tool can write files
  Write = 2,
  // This tool can perform arbitrary action
  Arbitrary = 3,
}

export type ToolInfo = {
  name: string
  description: string
  parameters: ToolParameter[]
  examples?: ToolExample[]
  permissionLevel: PermissionLevel
}

export type FullToolInfo = ToolInfo & {
  handler: ToolHandler<ToolInfo, any>
  isAvailable: (provider: any) => boolean
}

export enum ToolResponseType {
  Reply = 'Reply',
  Exit = 'Exit',
  Invalid = 'Invalid',
  Error = 'Error',
  Interrupted = 'Interrupted',
  HandOver = 'HandOver',
  Delegate = 'Delegate',
  Pause = 'Pause',
}

// Reply to the tool use
export type ToolResponseReply = {
  type: ToolResponseType.Reply
  message: string
}

// Should end the message thread
// e.g. task completed successfully
export type ToolResponseExit = {
  type: ToolResponseType.Exit
  message: string
  object?: any
}

// The tool arguments are invalid
export type ToolResponseInvalid = {
  type: ToolResponseType.Invalid
  message: string
}

// Some error occurred when executing the tool
// e.g. network request error, IO error
export type ToolResponseError = {
  type: ToolResponseType.Error
  message: string
  // If true, the tool can be retried
  // e.g. network request error are generally retryable
  // but IO errors are not
  canRetry?: boolean
}

// The tool execution was interrupted
// e.g. user cancelled the tool execution
// or some security policy was violated
export type ToolResponseInterrupted = {
  type: ToolResponseType.Interrupted
  message: string
}

// Hand over the task to another agent
// e.g. hand over a coding task to the coder agent
export type ToolResponseHandOver = {
  type: ToolResponseType.HandOver
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

// Delegate the task to another agent
export type ToolResponseDelegate = {
  type: ToolResponseType.Delegate
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

export type ToolResponsePause = {
  type: ToolResponseType.Pause
  object: any
}

export type ToolResponse =
  | ToolResponseReply
  | ToolResponseExit
  | ToolResponseInvalid
  | ToolResponseError
  | ToolResponseInterrupted
  | ToolResponseHandOver
  | ToolResponseDelegate
  | ToolResponsePause

export type ToolHandler<T extends ToolInfo, P> = (
  provider: P,
  args: Partial<Record<T['parameters'][number]['name'], ToolParameterValue>>,
) => Promise<ToolResponse>
