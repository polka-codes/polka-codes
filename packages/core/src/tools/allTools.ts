export { default as askFollowupQuestion } from './askFollowupQuestion'
export { default as attemptCompletion } from './attemptCompletion'
export { default as executeCommand } from './executeCommand'
export { default as listCodeDefinitionNames } from './listCodeDefinitionNames'
export { default as listFiles } from './listFiles'
export { default as readFile } from './readFile'
export { default as replaceInFile } from './replaceInFile'
export { default as searchFiles } from './searchFiles'
export { default as writeToFile } from './writeToFile'
export { default as handOver } from './handOver'
export { default as removeFile } from './removeFile'
export { default as renameFile } from './renameFile'
