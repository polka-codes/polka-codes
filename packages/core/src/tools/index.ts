export * from './provider'
export * from './allTools'
export { default as replaceInFile } from './replaceInFile' // not included in allTools

export * as allTools from './allTools'
