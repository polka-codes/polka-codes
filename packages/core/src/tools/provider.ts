export type FilesystemProvider = {
  readFile?: (path: string) => Promise<string>
  writeFile?: (path: string, content: string) => Promise<void>
  removeFile?: (path: string) => Promise<void>
  listFiles?: (path: string, recursive: boolean, maxCount: number) => Promise<[string[], boolean]>
  searchFiles?: (path: string, regex: string, filePattern: string) => Promise<string[]>
  listCodeDefinitionNames?: (path: string) => Promise<string[]>
}

export type CommandProvider = {
  executeCommand?: (command: string, needApprove: boolean) => Promise<{ stdout: string; stderr: string; exitCode: number }>
}

export type InteractionProvider = {
  askFollowupQuestion?: (question: string, options: string[]) => Promise<string>
  attemptCompletion?: (result: string) => Promise<string | undefined>
}

export type ToolProvider = FilesystemProvider & CommandProvider & InteractionProvider

export class MockProvider implements ToolProvider {
  async readFile(path: string): Promise<string> {
    return 'mock content'
  }

  async writeFile(path: string, content: string): Promise<void> {
    return
  }

  async removeFile(path: string): Promise<void> {
    return
  }

  async listFiles(path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> {
    return [['mock-file.txt'], false]
  }

  async searchFiles(path: string, regex: string, filePattern: string): Promise<string[]> {
    return ['mock-file.txt']
  }

  async listCodeDefinitionNames(path: string): Promise<string[]> {
    return ['mockDefinition']
  }

  async executeCommand(command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return { stdout: 'mock output', stderr: '', exitCode: 0 }
  }

  async askFollowupQuestion(question: string, options?: string[]): Promise<string> {
    return 'mock answer'
  }

  async attemptCompletion(result: string): Promise<string | undefined> {
    return 'mock completion'
  }
}
