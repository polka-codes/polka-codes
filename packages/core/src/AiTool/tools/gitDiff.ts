// generated by polka.codes
import { z } from 'zod'
import { type FullToolInfoV2, PermissionLevel, type ToolHandler, type ToolInfoV2, ToolResponseType } from '../../tool'
import type { CommandProvider } from '../../tools'

export const toolInfo = {
  name: 'git_diff',
  description:
    'Get the git diff for the current repository. Can be used to get staged changes, unstaged changes, or changes between commits. By default, it returns unstaged changes.',
  parameters: z.object({
    staged: z
      .preprocess((val) => {
        if (typeof val === 'string') {
          const lower = val.toLowerCase()
          if (lower === 'false') return false
          if (lower === 'true') return true
        }
        return val
      }, z.boolean().optional().default(false))
      .describe('Get staged changes instead of unstaged changes.'),
    commitRange: z.string().optional().describe('The commit range to get the diff for (e.g., "main...HEAD").'),
    file: z.string().optional().describe('Get the diff for a specific file.'),
  }),
  permissionLevel: PermissionLevel.Read,
} as const satisfies ToolInfoV2

export const handler: ToolHandler<typeof toolInfo, CommandProvider> = async (provider, args) => {
  if (!provider.executeCommand) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to execute command. Abort.',
    }
  }

  const { staged, file, commitRange } = toolInfo.parameters.parse(args)

  const commandParts = ['git', 'diff', '--no-color']
  if (staged) {
    commandParts.push('--staged')
  }
  if (commitRange) {
    commandParts.push(commitRange)
  }
  if (file) {
    commandParts.push('--', file)
  }

  const command = commandParts.join(' ')
  try {
    const result = await provider.executeCommand(command, false)
    if (result.exitCode === 0) {
      if (!result.stdout.trim()) {
        return {
          type: ToolResponseType.Reply,
          message: 'No diff found.',
        }
      }
      return {
        type: ToolResponseType.Reply,
        message: `<diff file="${file ?? 'all'}">\n${result.stdout}\n</diff>`,
      }
    }
    return {
      type: ToolResponseType.Error,
      message: `\`${command}\` exited with code ${result.exitCode}:\n${result.stderr}`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Error,
      message: error instanceof Error ? error.message : String(error),
    }
  }
}

export const isAvailable = (provider: CommandProvider): boolean => {
  return !!provider.executeCommand
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfoV2
