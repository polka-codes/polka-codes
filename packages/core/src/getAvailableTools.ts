import type { FullToolInfo, PermissionLevel } from './tool'
import { askFollowupQuestion, delegate, handOver } from './tools'

/**
 * Get the available tools based on the provider and agent availability
 * @param provider The provider to use
 * @param allTools All possible tools
 * @param hasAgent Whether the agent has agents
 * @param permissionLevel The permission level of the agent. Tool requires higher permission level will be disabled
 * @returns The available tools
 */
export const getAvailableTools = ({
  provider,
  allTools,
  hasAgent,
  permissionLevel,
  interactive,
}: { provider: any; allTools: FullToolInfo[]; hasAgent: boolean; permissionLevel: PermissionLevel; interactive: boolean }) => {
  const filteredTools = interactive ? allTools : allTools.filter((tool) => tool.name !== askFollowupQuestion.name)
  const tools: FullToolInfo[] = []
  for (const tool of filteredTools) {
    // disable agent tools if no agents available
    if (!hasAgent) {
      switch (tool.name) {
        case handOver.name:
        case delegate.name:
          continue
      }
    }
    if (tool.isAvailable(provider) && tool.permissionLevel <= permissionLevel) {
      tools.push(tool)
    }
  }
  return tools
}
