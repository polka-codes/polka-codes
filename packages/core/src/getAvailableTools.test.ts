/**
 * @file Test file for getAvailableTools function
 * Generated by polka.codes
 */

import { describe, expect, it } from 'bun:test'
import { z } from 'zod'
import { getAvailableTools } from './getAvailableTools'
import type { FullToolInfo, FullToolInfoV2, PermissionLevel } from './tool'
import { ToolResponseType } from './tool'
import { toToolInfoV1 } from './tool-v1-compat'
import { delegate, handOver } from './tools'
import { MockProvider } from './tools/provider'

describe('getAvailableTools', () => {
  const mockProvider = new MockProvider()

  // Mock tools for testing
  const createMockTool = (name: string, isAvailable = true, permissionLevel: PermissionLevel = 1): FullToolInfoV2 => ({
    name,
    description: `Mock tool ${name}`,
    parameters: z.object({}),
    handler: async () => ({
      type: ToolResponseType.Reply,
      message: 'Mock response',
    }),
    isAvailable: () => isAvailable,
    permissionLevel,
  })

  const mockTools: FullToolInfo[] = [
    toToolInfoV1(delegate),
    toToolInfoV1(handOver),
    toToolInfoV1(createMockTool('tool1', true, 1)),
    toToolInfoV1(createMockTool('tool2', false, 1)),
    toToolInfoV1(createMockTool('tool3', true, 2)),
    toToolInfoV1(createMockTool('tool4', true, 3)),
  ]

  it('should filter out agent tools when no agents available', () => {
    const result = getAvailableTools({
      provider: mockProvider,
      allTools: mockTools,
      hasAgent: false,
      permissionLevel: 3,
      interactive: true,
    })

    // Should not include delegate and handOver
    expect(result.some((tool) => tool.name === delegate.name)).toBe(false)
    expect(result.some((tool) => tool.name === handOver.name)).toBe(false)

    // Should include other available tools
    expect(result.some((tool) => tool.name === 'tool1')).toBe(true)

    expect(result).toMatchSnapshot()
  })

  it('should include agent tools when agents are available', () => {
    const result = getAvailableTools({ provider: mockProvider, allTools: mockTools, hasAgent: true, permissionLevel: 3, interactive: true })

    // Should include delegate and handOver
    expect(result.some((tool) => tool.name === delegate.name)).toBe(true)
    expect(result.some((tool) => tool.name === handOver.name)).toBe(true)

    expect(result).toMatchSnapshot()
  })

  it('should filter based on provider availability', () => {
    const result = getAvailableTools({ provider: mockProvider, allTools: mockTools, hasAgent: true, permissionLevel: 3, interactive: true })

    // Should include available tools
    expect(result.some((tool) => tool.name === 'tool1')).toBe(true)

    // Should exclude unavailable tools
    expect(result.some((tool) => tool.name === 'tool2')).toBe(false)

    expect(result).toMatchSnapshot()
  })

  it('should filter based on permission level', () => {
    const result = getAvailableTools({ provider: mockProvider, allTools: mockTools, hasAgent: true, permissionLevel: 2, interactive: true })

    // Should include tools with permissionLevel <= 2
    expect(result.some((tool) => tool.name === 'tool1')).toBe(true)
    expect(result.some((tool) => tool.name === 'tool3')).toBe(true)

    // Should exclude tools with permissionLevel > 2
    expect(result.some((tool) => tool.name === 'tool4')).toBe(false)

    expect(result).toMatchSnapshot()
  })

  it('should correctly apply all filters together', () => {
    const result = getAvailableTools({
      provider: mockProvider,
      allTools: mockTools,
      hasAgent: false,
      permissionLevel: 1,
      interactive: true,
    })

    // Should exclude agent tools (no agents)
    expect(result.some((tool) => tool.name === delegate.name)).toBe(false)
    expect(result.some((tool) => tool.name === handOver.name)).toBe(false)

    // Should exclude unavailable tools
    expect(result.some((tool) => tool.name === 'tool2')).toBe(false)

    // Should exclude tools with higher permission level
    expect(result.some((tool) => tool.name === 'tool3')).toBe(false)
    expect(result.some((tool) => tool.name === 'tool4')).toBe(false)

    // Should only include available tools with correct permission level
    expect(result.some((tool) => tool.name === 'tool1')).toBe(true)

    expect(result).toMatchSnapshot()
  })
})
