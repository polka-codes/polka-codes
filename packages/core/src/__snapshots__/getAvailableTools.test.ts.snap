// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`getAvailableTools should filter out agent tools when no agents available 1`] = `
[
  {
    "description": "Mock tool tool1",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool1",
    "parameters": [],
    "permissionLevel": 1,
  },
  {
    "description": "Mock tool tool3",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool3",
    "parameters": [],
    "permissionLevel": 2,
  },
  {
    "description": "Mock tool tool4",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool4",
    "parameters": [],
    "permissionLevel": 3,
  },
]
`;

exports[`getAvailableTools should include agent tools when agents are available 1`] = `
[
  {
    "description": "Temporarily delegate a task to another agent and receive the result back. This tool MUST NOT to be used with any other tool.",
    "examples": [
      {
        "description": "Delegate a code analysis task to the analyzer agent",
        "input": {
          "agentName": "analyzer",
          "context": "Need to understand the security implications of the current auth system",
          "files": "src/auth/login.ts,src/auth/types.ts",
          "task": "Analyze the authentication implementation",
        },
      },
    ],
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "delegate",
    "parameters": [
      {
        "description": "The name of the agent to delegate the task to",
        "name": "agentName",
        "required": true,
        "usageValue": "Name of the target agent",
      },
      {
        "description": "The task to be completed by the target agent",
        "name": "task",
        "required": true,
        "usageValue": "Task description",
      },
      {
        "description": "The context information for the task",
        "name": "context",
        "required": true,
        "usageValue": "Context information",
      },
      {
        "description": "The files relevant to the task. Comma separated paths",
        "name": "files",
        "required": false,
        "usageValue": "Relevant files",
      },
    ],
    "permissionLevel": 0,
  },
  {
    "description": "Hand over the current task to another agent to complete. This tool MUST NOT to be used with any other tool.",
    "examples": [
      {
        "description": "Hand over a coding task to the coder agent",
        "input": {
          "agentName": "coder",
          "context": "We need a secure login system with email and password",
          "files": "src/auth/login.ts,src/auth/types.ts",
          "task": "Implement the login feature",
        },
      },
    ],
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "hand_over",
    "parameters": [
      {
        "description": "The name of the agent to hand over the task to",
        "name": "agentName",
        "required": true,
        "usageValue": "Name of the target agent",
      },
      {
        "description": "The task to be completed by the target agent",
        "name": "task",
        "required": true,
        "usageValue": "Task description",
      },
      {
        "description": "The context information for the task",
        "name": "context",
        "required": true,
        "usageValue": "Context information",
      },
      {
        "description": "The files relevant to the task. Comma separated paths",
        "name": "files",
        "required": false,
        "usageValue": "Relevant files",
      },
    ],
    "permissionLevel": 0,
  },
  {
    "description": "Mock tool tool1",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool1",
    "parameters": [],
    "permissionLevel": 1,
  },
  {
    "description": "Mock tool tool3",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool3",
    "parameters": [],
    "permissionLevel": 2,
  },
  {
    "description": "Mock tool tool4",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool4",
    "parameters": [],
    "permissionLevel": 3,
  },
]
`;

exports[`getAvailableTools should filter based on provider availability 1`] = `
[
  {
    "description": "Temporarily delegate a task to another agent and receive the result back. This tool MUST NOT to be used with any other tool.",
    "examples": [
      {
        "description": "Delegate a code analysis task to the analyzer agent",
        "input": {
          "agentName": "analyzer",
          "context": "Need to understand the security implications of the current auth system",
          "files": "src/auth/login.ts,src/auth/types.ts",
          "task": "Analyze the authentication implementation",
        },
      },
    ],
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "delegate",
    "parameters": [
      {
        "description": "The name of the agent to delegate the task to",
        "name": "agentName",
        "required": true,
        "usageValue": "Name of the target agent",
      },
      {
        "description": "The task to be completed by the target agent",
        "name": "task",
        "required": true,
        "usageValue": "Task description",
      },
      {
        "description": "The context information for the task",
        "name": "context",
        "required": true,
        "usageValue": "Context information",
      },
      {
        "description": "The files relevant to the task. Comma separated paths",
        "name": "files",
        "required": false,
        "usageValue": "Relevant files",
      },
    ],
    "permissionLevel": 0,
  },
  {
    "description": "Hand over the current task to another agent to complete. This tool MUST NOT to be used with any other tool.",
    "examples": [
      {
        "description": "Hand over a coding task to the coder agent",
        "input": {
          "agentName": "coder",
          "context": "We need a secure login system with email and password",
          "files": "src/auth/login.ts,src/auth/types.ts",
          "task": "Implement the login feature",
        },
      },
    ],
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "hand_over",
    "parameters": [
      {
        "description": "The name of the agent to hand over the task to",
        "name": "agentName",
        "required": true,
        "usageValue": "Name of the target agent",
      },
      {
        "description": "The task to be completed by the target agent",
        "name": "task",
        "required": true,
        "usageValue": "Task description",
      },
      {
        "description": "The context information for the task",
        "name": "context",
        "required": true,
        "usageValue": "Context information",
      },
      {
        "description": "The files relevant to the task. Comma separated paths",
        "name": "files",
        "required": false,
        "usageValue": "Relevant files",
      },
    ],
    "permissionLevel": 0,
  },
  {
    "description": "Mock tool tool1",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool1",
    "parameters": [],
    "permissionLevel": 1,
  },
  {
    "description": "Mock tool tool3",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool3",
    "parameters": [],
    "permissionLevel": 2,
  },
  {
    "description": "Mock tool tool4",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool4",
    "parameters": [],
    "permissionLevel": 3,
  },
]
`;

exports[`getAvailableTools should filter based on permission level 1`] = `
[
  {
    "description": "Temporarily delegate a task to another agent and receive the result back. This tool MUST NOT to be used with any other tool.",
    "examples": [
      {
        "description": "Delegate a code analysis task to the analyzer agent",
        "input": {
          "agentName": "analyzer",
          "context": "Need to understand the security implications of the current auth system",
          "files": "src/auth/login.ts,src/auth/types.ts",
          "task": "Analyze the authentication implementation",
        },
      },
    ],
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "delegate",
    "parameters": [
      {
        "description": "The name of the agent to delegate the task to",
        "name": "agentName",
        "required": true,
        "usageValue": "Name of the target agent",
      },
      {
        "description": "The task to be completed by the target agent",
        "name": "task",
        "required": true,
        "usageValue": "Task description",
      },
      {
        "description": "The context information for the task",
        "name": "context",
        "required": true,
        "usageValue": "Context information",
      },
      {
        "description": "The files relevant to the task. Comma separated paths",
        "name": "files",
        "required": false,
        "usageValue": "Relevant files",
      },
    ],
    "permissionLevel": 0,
  },
  {
    "description": "Hand over the current task to another agent to complete. This tool MUST NOT to be used with any other tool.",
    "examples": [
      {
        "description": "Hand over a coding task to the coder agent",
        "input": {
          "agentName": "coder",
          "context": "We need a secure login system with email and password",
          "files": "src/auth/login.ts,src/auth/types.ts",
          "task": "Implement the login feature",
        },
      },
    ],
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "hand_over",
    "parameters": [
      {
        "description": "The name of the agent to hand over the task to",
        "name": "agentName",
        "required": true,
        "usageValue": "Name of the target agent",
      },
      {
        "description": "The task to be completed by the target agent",
        "name": "task",
        "required": true,
        "usageValue": "Task description",
      },
      {
        "description": "The context information for the task",
        "name": "context",
        "required": true,
        "usageValue": "Context information",
      },
      {
        "description": "The files relevant to the task. Comma separated paths",
        "name": "files",
        "required": false,
        "usageValue": "Relevant files",
      },
    ],
    "permissionLevel": 0,
  },
  {
    "description": "Mock tool tool1",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool1",
    "parameters": [],
    "permissionLevel": 1,
  },
  {
    "description": "Mock tool tool3",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool3",
    "parameters": [],
    "permissionLevel": 2,
  },
]
`;

exports[`getAvailableTools should correctly apply all filters together 1`] = `
[
  {
    "description": "Mock tool tool1",
    "handler": [Function: AsyncFunction],
    "isAvailable": [Function],
    "name": "tool1",
    "parameters": [],
    "permissionLevel": 1,
  },
]
`;
